File: shanoir-ng-front/src/app/study-cards/shared/quality-card.service.ts
/**
 * Shanoir NG - Import, manage and share neuroimaging data
 * Copyright (C) 2009-2019 Inria - https://www.inria.fr/
 * Contact us on https://project.inria.fr/shanoir/
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/gpl-3.0.html
 */

import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { firstValueFrom } from 'rxjs';

import { EntityService } from '../../shared/components/entity/entity.abstract.service';
import * as AppUtils from '../../utils/app.utils';

import { QualityCardDTOService } from './quality-card.dto';
import { QualityCardDTO } from './quality-card.dto.model';
import { QualityCard } from './quality-card.model';

export class Interval {
    constructor(public from: number, public to: number) {}
};

@Injectable()
export class QualityCardService extends EntityService<QualityCard> {

    API_URL = AppUtils.BACKEND_API_QUALITY_CARD_URL;

    private qualityCardDTOService: QualityCardDTOService = inject(QualityCardDTOService);
    
    constructor(protected http: HttpClient) {
        super(http)
    }

    getEntityInstance() { return new QualityCard(); }

    getAllForStudy(studyId: number): Promise<QualityCard[]> {
        return firstValueFrom(this.http.get<any[]>(this.API_URL + '/byStudy/' + studyId))
            .then(this.mapEntityList);
    }

    protected mapEntity = (dto: QualityCardDTO, result?: QualityCard): Promise<QualityCard> => {
        if (result == undefined) result = this.getEntityInstance();
        return this.qualityCardDTOService.toEntity(dto, result);
    }

    protected mapEntityList = (dtos: QualityCardDTO[], result?: QualityCard[]): Promise<QualityCard[]> => {
        if (result == undefined) result = [];
        if (dtos) return this.qualityCardDTOService.toEntityList(dtos, result);
    }

    public stringify(entity: QualityCard) {
        const dto = new QualityCardDTO(entity);
        return JSON.stringify(dto, (key, value) => {
            return this.customReplacer(key, value, dto);
        });
    }

    applyOnStudy(qualityCardId: number): Promise<any> {
        return firstValueFrom(this.http.get<any[]>(this.API_URL + '/apply/' + qualityCardId));
    }

    testOnStudy(qualityCardId: number, start?: number, stop?: number): Promise<any> {
        return firstValueFrom(this.http.get<any[]>(this.API_URL + '/test/' + qualityCardId 
            + (start != null && start != undefined && stop != null && stop != undefined
                ? '/' + (start - 1) + '/' + stop : '')));
    }

    testOnExamination(qualityCardId: number, examinationId: number): Promise<any> {
        return firstValueFrom(this.http.get<any[]>(this.API_URL + '/test/' + qualityCardId + '/exam/' + examinationId));
    }

}

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/QualityCardService.java
/**
 * Shanoir NG - Import, manage and share neuroimaging data
 * Copyright (C) 2009-2019 Inria - https://www.inria.fr/
 * Contact us on https://project.inria.fr/shanoir/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/gpl-3.0.html
 */

package org.shanoir.ng.studycard.service;

import java.util.List;

import org.shanoir.ng.datasetacquisition.model.DatasetAcquisition;
import org.shanoir.ng.download.AcquisitionAttributes;
import org.shanoir.ng.shared.exception.EntityNotFoundException;
import org.shanoir.ng.shared.exception.PacsException;
import org.shanoir.ng.shared.exception.RestServiceException;
import org.shanoir.ng.shared.exception.ShanoirException;
import org.shanoir.ng.studycard.dto.QualityCardResult;
import org.shanoir.ng.studycard.model.QualityCard;
import org.springframework.validation.BindingResult;

public interface QualityCardService {

    /**
     * Validate a quality card
     *
     * @param qualityCard
     * @param result
     * @throws RestServiceException
     */
    void validate(QualityCard qualityCard, BindingResult result) throws RestServiceException;

    /**
     * Update an existing quality card
     *
     * @param qualityCard
     */
    void update(QualityCard qualityCard) throws EntityNotFoundException;

    /**
     * Nulls out the ids of the quality card's rules and their nested conditions, so that saving it
     * (creation, or duplication of an existing quality card) inserts fresh rows instead of failing on
     * stale/foreign ids carried over from the client payload.
     *
     * @param qualityCard
     */
    void resetIdsForFreshInsert(QualityCard qualityCard);

    /**
     * Quality cards for quality control: apply on entire study.
     *
     * @param qualityCard
     * @param updateTags for testing or for real apply
     */
    QualityCardResult applyQualityCardOnStudy(QualityCard qualityCard, boolean updateTags) throws PacsException;

    /**
     * Quality cards for quality control: apply on a sample of the study's examinations only
     * (used by the "test on sample" option for studies with too many examinations to test in full).
     *
     * @param qualityCard
     * @param updateTags for testing or for real apply
     * @param from index of the first examination to process, or null to process all
     * @param to index of the last examination to process, or null to process all
     */
    QualityCardResult applyQualityCardOnStudy(QualityCard qualityCard, boolean updateTags, Integer from, Integer to) throws PacsException;

    QualityCardResult checkQuality(DatasetAcquisition datasetAcquisition, AcquisitionAttributes<?> acquisitionAttributes, List<QualityCard> qualityCards) throws ShanoirException;
}
